File: lintcode/sqrtx.py
#!/usr/bin/env python
# encoding: utf-8

# Question: http://www.lintcode.com/en/problem/sqrtx/

class Solution:
    """
    @param x: An integer
    @return: The sqrt of x
    """
    def sqrt(self, x):
        # write your code here
        if x < 0:
            return -1
        if x == 0:
            return 0

        l, h = 1, x
        while l+1 < h:
            mid = (l+h)/2
            p = pow(mid, 2)
            if p == x:
                return mid
            elif p > x:
                h = mid
            else:
                l = mid

        return l

if __name__ == '__main__':
    s = Solution()
    print s.sqrt(0)
    print s.sqrt(1)
    print s.sqrt(2)
    print s.sqrt(9)
    print s.sqrt(10)
    print s.sqrt(11)
    print s.sqrt(12)

File: leetcode-2/sqrtx.cc
// Question: https://leetcode.com/problems/sqrtx/

class Solution {
public:
    int mySqrt(int x) {
        if (x <= 0) {
            return 0;
        }
        int l = 1, r = x;
        while (l <= r) {
            int m = l + (r-l)/2;
            int t = x/m;
            if (t == m) {
                return m;
            } else if (t > m) {
                l = m;
            } else {
                r = m - 1;
            }
        }
        return l;
    }
};
